src/lib/Types/Core.hs

```diff
@@ -2432,6 +2445,55 @@ instance Store IxMethod
 instance Store ParamRole
 instance Store (SpecializedDictDef n)
 
+-- === substituting Imp names with CAtoms
+
+type IAtomSubstVal = SubstVal ImpNameC (Atom SimpToImpIR)
+instance SubstE IAtomSubstVal (Atom SimpToImpIR) where
+  substE (scope, env) atom = case fromE atom of
+    Case0 rest -> (toE . Case0) $ substE (scope, env) rest
+    Case1 rest -> (toE . Case1) $ substE (scope, env) rest
+    Case2 rest -> (toE . Case2) $ substE (scope, env) rest
+    Case3 rest -> (toE . Case3) $ substE (scope, env) rest
+    Case4 rest -> (toE . Case4) $ substE (scope, env) rest
+    Case5 specialCase -> case specialCase of
+      Case0 rest -> toE $ Case5 $ Case0 $ substE (scope, env) rest
+      Case1 rest -> toE $ Case5 $ Case1 $ substE (scope, env) rest
+      -- AtomicIVar
+      Case2 (LiftE ReifiedSat `PairE` LeftE v `PairE` LiftE ty) -> do
+        case env ! v of
+          Rename v' -> AtomicIVar (LeftE v') ty
+          SubstVal x -> x
+      Case2 (LiftE ReifiedSat `PairE` RightE v `PairE` LiftE ty) -> do
+        case env ! v of
+          Rename v' -> AtomicIVar (RightE v') ty
+      _ -> error "impossible"
+    Case6 rest -> (toE . Case6) $ substE (scope, env) rest
+    Case7 rest -> (toE . Case7) $ substE (scope, env) rest
+
+instance (SubstE IAtomSubstVal ann, SinkableE ann) => SubstB IAtomSubstVal (NonDepNest SimpToImpIR ann)
+instance SubstE IAtomSubstVal (BoxPtr  SimpToImpIR)
+instance SubstE IAtomSubstVal (Expr    SimpToImpIR)
+instance SubstE IAtomSubstVal (Block   SimpToImpIR)
+instance SubstE IAtomSubstVal (DeclBinding SimpToImpIR)
+instance SubstB IAtomSubstVal (Decl        SimpToImpIR)
+instance SubstB IAtomSubstVal (LamBinder   SimpToImpIR)
+instance SubstE IAtomSubstVal (EffectP Name)
+instance SubstE IAtomSubstVal (EffectRowP Name)
+instance SubstE IAtomSubstVal (LamExpr SimpToImpIR)
+instance SubstE IAtomSubstVal (ExtLabeledItemsE (Type SimpToImpIR) UnitE)
+instance SubstE IAtomSubstVal (ExtLabeledItemsE (Type SimpToImpIR) (AtomName SimpToImpIR))
+instance SubstE IAtomSubstVal (FieldRowElem  SimpToImpIR)
+instance SubstE IAtomSubstVal (FieldRowElems SimpToImpIR)
+instance SubstE IAtomSubstVal (DepPairType SimpToImpIR)
+instance SubstE IAtomSubstVal (DataDefParams SimpToImpIR)
+instance SubstE IAtomSubstVal (PiType SimpToImpIR)
+instance SubstB IAtomSubstVal (PiBinder SimpToImpIR)
+instance SubstE IAtomSubstVal (DictExpr SimpToImpIR)
+instance SubstE IAtomSubstVal (DictType SimpToImpIR)
+instance SubstE IAtomSubstVal (TabLamExpr SimpToImpIR)
+instance SubstE IAtomSubstVal (TabPiType SimpToImpIR)
+instance SubstE IAtomSubstVal (IxType SimpToImpIR)
```

Comment:
Oof that's a lot of extra code to generate. Why do we need this?

Reply:
We use it in two places. (1) the atom reconstruction at the top level. This will go away once we can work with canonical atoms, and (2) we rename imp names to core names before calling into algebra because otherwise we would have a mix. I'd like to delete this soon, but I don't know how to solve (2). Any ideas?

Reply:
Well the big part that I don't understand is: why do we end up with this mix of IRs in algebra? Is it because some names might refer to top-level values, while some might be defined locally in the computation? Aren't we inlining all top level values anyway?

Reply:
Right, top-level names aren't the issue. The reason we have a mix is that there are Imp names due to ordinary local Imp bindings (e.g. some runtime computed `n`), but there are also Atom names due to index binders in the dependent table types. This is because we do the algebra stuff on the core/simplified IR. If we switched to doing it in Imp IR I think we'd be fine.